pinot-segment-local/src/main/java/org/apache/pinot/segment/local/function/GroovyFunctionEvaluator.java

```diff
@@ -72,13 +83,65 @@ public GroovyFunctionEvaluator(String closure) {
     }
     _numArguments = _arguments.size();
     _binding = new Binding();
-    _script = new GroovyShell(_binding).parse(matcher.group(SCRIPT_GROUP_NAME));
+    final String scriptText = matcher.group(SCRIPT_GROUP_NAME);
+
+    final GroovyStaticAnalyzerConfig groovyStaticAnalyzerConfig = getConfig();
+    _script = createSafeShell(_binding, groovyStaticAnalyzerConfig).parse(scriptText);
   }
 
   public static String getGroovyExpressionPrefix() {
     return GROOVY_EXPRESSION_PREFIX;
   }
 
+  /**
+   * This will create a Groovy Shell that is configured with static syntax analysis. This static syntax analysis
+   * will that any script which is run is restricted to a specific list of allowed operations, thus making it harder
+   * to execute malicious code.
+   *
+   * @param binding Binding instance to be used by Groovy Shell.
+   * @param groovyConfig GroovyStaticAnalyzerConfig instance to be used for static syntax analysis.
+   * @return GroovyShell instance with static syntax analysis.
+   */
+  private GroovyShell createSafeShell(Binding binding, GroovyStaticAnalyzerConfig groovyConfig) {
```

Comment:
Do you anticipate performance overhead for the safe shell?

Reply:
No. There should not be any performance impact. The `CompilerConfiguration` is created only once and reused. It is updated everytime static analyzer config is updated.